README.md
# Scrollable Bottom Sheet

This is a temporary workaround to achieve Scrollable Persistent Bottom Sheet that i created.

## Installation

First, add `scrollable_bottom_sheet` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

```
scrollable_bottom_sheet: ^0.1.0+2
```

## Example
```
class _BottomSheetDemoState extends State<BottomSheetDemo>
    with TickerProviderStateMixin {
  bool _bottomSheetActive = false;
  String _currentState = "initial";
  String _currentDirection = "up";

  void _showMessage(BuildContext context) {
    showDialog<void>(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          content: const Text('You tapped the floating action button.'),
          actions: <Widget>[
            FlatButton(
                onPressed: () {
                  Navigator.pop(context);
                },
                child: const Text('OK'))
          ],
        );
      },
    );
  }

  final GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey<ScaffoldState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _scaffoldKey,
      appBar: AppBar(
        title: const Text('Persistent bottom sheet'),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _showMessage(context);
        },
        backgroundColor: Colors.redAccent,
        child: const Icon(
          Icons.add,
          semanticLabel: 'Add',
        ),
      ),
      body: Builder(
        builder: (BuildContext context) {
          return Stack(children: [
            Center(
                child: RaisedButton(
                    onPressed: _bottomSheetActive
                        ? null
                        : () {
                            setState(() {
                              //disable button
                              _bottomSheetActive = true;
                            });
                            _showBottomSheet(context);
                          },
                    child: const Text('Show bottom sheet'))),
          ]);
        },
      ),
    );
  }

  Widget _bottomSheetBuilder(BuildContext context) {
    final key = new GlobalKey<ScrollableBottomSheetState>();
    final ThemeData themeData = Theme.of(context);
    AnimationController animationController = AnimationController(vsync: this);

    return Stack(children: [
      ScrollableBottomSheet(
        key: key,
        halfHeight: 250.0,
        minimumHeight: 50.0,
        autoPop: false,
        scrollTo: ScrollState.minimum,
        snapAbove: false,
        snapBelow: false,
        callback: (state) {
          if (state == ScrollState.minimum) {
            _currentState = "minimum";
            _currentDirection = "up";
          } else if (state == ScrollState.half) {
            if (_currentState == "minimum") {
              _currentDirection = "up";
            } else {
              _currentDirection = "down";
            }
            _currentState = "half";
          } else {
            _currentState = "full";
            _currentDirection = "down";
          }
        },
        child: Container(
            color: Colors.greenAccent,
            margin: EdgeInsets.only(bottom: 50.0),
            child: Padding(
                padding: const EdgeInsets.all(32.0),
                child: Column(children: [
                  InkWell(
                    child: Container(color: Colors.red, height: 57.0),
                    onTap: () {
                      key.currentState.animateToZero(context, willPop: true);
                    },
                  ),
                  Text(
                      'This is a Material persistent bottom sheet. Drag downwards to dismiss it.',
                      textAlign: TextAlign.center,
                      style: TextStyle(
                          color: themeData.accentColor, fontSize: 24.0)),
                  Column(
                      children: List.generate(100, (index) {
                    return Text("Text $index");
                  }))
                ]))),
      ),
      Positioned(
          bottom: 0.0,
          left: 0.0,
          right: 0.0,
          height: 50.0,
          child: Material(
            elevation: 15.0,
            child: IconButton(
                icon: Icon(Icons.location_on),
                onPressed: () {
                  if (_currentState == "half") {
                    if (_currentDirection == "up") {
                      key.currentState.animateToFull(context);
                    } else {
                      key.currentState.animateToMinimum(context);
                    }
                  } else {
                    key.currentState.animateToHalf(context);
                  }
                }),
          ))
    ]);
  }

  _showBottomSheet(BuildContext context) {
    showBottomSheet<void>(context: context, builder: _bottomSheetBuilder)
        .closed
        .whenComplete(() {
      if (mounted) {
        setState(() {
          // re-enable the button
          _bottomSheetActive = false;
        });
      }
    });
  }
}
```


CHANGELOG.md
## 0.0.1

* Initial release.

## 0.0.2

* Bug fixes and snap feature added

## 0.0.3

* Minimum feature and animation control from the outside added

## 0.0.4

* Bug Fixes

## 0.0.5

* Bug Fixes, add callback feature, and renaming some

## 0.0.6

* Fling scroll feature added

## 0.0.7

* Bug Fixes

## 0.0.8

* Scroll smoothness improved, add new feature measure minimum and half height by content

## 0.0.9

* Minor bug Fixes

## 0.1.0

* Change communication type between parent and bottom sheet from GlobalKey to ScrollableController

## 0.1.0+1

* Change ScrollableBottomSheetState to be private since there's no need to create GlobalKey anymore

## 0.1.0+2

* Fix error with no controller in ScrollableBottomSheet

lib/scrollable_bottom_sheet.dart
import 'dart:math' as math;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:scrollable_bottom_sheet/scrollable_controller.dart';

typedef AnimationCallback = void Function(double value);
typedef StateCallback = void Function(ScrollState state);

enum ScrollDirection { none, up, down }
enum ScrollState { full, half, minimum }

class ScrollableBottomSheetByContent extends StatefulWidget {
  final Widget header;
  final Widget content;
  final ScrollableController controller;
  final StateCallback callback;
  final bool snapAbove;
  final bool snapBelow;
  final bool autoPop;
  final ScrollState scrollTo;

  ScrollableBottomSheetByContent(this.header, this.content,
      {ScrollableController controller, bool snapAbove, bool snapBelow, bool autoPop, this.callback, ScrollState scrollTo})
      : this.controller = controller ?? ScrollableController(),
        this.snapAbove = snapAbove ?? true,
        this.snapBelow = snapBelow ?? true,
        this.autoPop = autoPop ?? true,
        this.scrollTo = scrollTo ?? ScrollState.minimum;

  @override
  State<StatefulWidget> createState() => _ScrollableBottomSheetByContentState();
}

class _ScrollableBottomSheetByContentState extends State<ScrollableBottomSheetByContent> {
  BuildContext _headerContext;
  BuildContext _contentContext;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      widget.controller.setMinimumHeight(_headerContext.size.height);
      widget.controller.setHalfHeight(_headerContext.size.height + _contentContext.size.height);

      if (widget.scrollTo == ScrollState.minimum) {
        widget.controller.animateToMinimum(context);
      } else if (widget.scrollTo == ScrollState.half) {
        widget.controller.animateToHalf(context);
      } else if (widget.scrollTo == ScrollState.full) {
        widget.controller.animateToFull(context);
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return ScrollableBottomSheet(
      controller: widget.controller,
      snapAbove: widget.snapAbove,
      snapBelow: widget.snapBelow,
      autoPop: widget.autoPop,
      callback: widget.callback,
      child: Column(children: [
        Builder(
          builder: (BuildContext context) {
            _headerContext = context;
            return widget.header;
          },
        ),
        Builder(
          builder: (BuildContext context) {
            _contentContext = context;
            return widget.content;
          },
        ),
      ]),
      halfHeight: 0.0,
    );
  }
}

class ScrollableBottomSheet extends StatefulWidget {
  /// We change from GlobalKey system to controller system
  /// Whenever parent wants to animate bottom sheet to certain position
  /// for example minimum / half / full, it is executed by this controller
  ///
  /// See [ScrollableInterface]
  final ScrollableController controller;

  /// The Bottom Sheet's height when first open
  /// Must not be null
  final double halfHeight;

  /// The content inside Bottom Sheet
  /// Must not be null
  final Widget child;

  /// If this true, if user drag Bottom Sheet in between [halfHeight] and
  /// maximum height, the Bottom Sheet will be snapped according to last scroll
  /// direction user made
  ///
  /// If this false, there will be no snapping automatically
  ///
  /// See [ScrollDirection], [minimumHeight]
  final bool snapAbove;

  /// If this true, if user drag Bottom Sheet in between [halfHeight] and
  /// [minimumHeight], the Bottom Sheet will be snapped according to last scroll
  /// direction user made
  ///
  /// If this false, there will be no snapping automatically
  ///
  /// See [ScrollDirection], [minimumHeight]
  final bool snapBelow;

  /// If this true, if user drag Bottom Sheet until [minimumHeight],
  /// Navigation.pop will be called automatically
  ///
  /// If this false, there will be no popping automatically
  ///
  /// See [minimumHeight]
  final bool autoPop;

  /// If this value is not null, this value will replace 0.0 as minimum height,
  /// and there will be no automatic popping when user scroll Bottom Sheet
  /// between [halfHeight] and [minimumHeight]
  ///
  /// You can pop it manually by using [Navigator]
  ///
  /// ```
  /// Navigator.pop(context);
  /// ```
  ///
  /// See [halfHeight]
  final double minimumHeight;

  /// Scroll State callback, it triggered every time bottom sheet reach for
  /// fullHeight / [halfHeight] / [minimumHeight]
  ///
  /// See [ScrollState], [halfHeight], [minimumHeight]
  final StateCallback callback;

  /// To determine which position the bottom sheet should be in the first push
  /// between fullHeight / [halfHeight] / [minimumHeight]
  ///
  /// See [ScrollState], [halfHeight], [minimumHeight]
  final ScrollState firstScrollTo;

  /// If this true, fullHeight won't exceed child's height
  ///
  /// If this false, it'll always scrollable to fullscreen height
  ///
  /// See [_fullHeight]
  final bool mayExceedChildHeight;

  ScrollableBottomSheet(
      {ScrollableController controller,
      @required this.halfHeight,
      @required this.child,
      bool snapAbove,
      bool snapBelow,
      bool autoPop,
      double minimumHeight,
      this.callback,
      ScrollState scrollTo,
      bool mayExceedChildHeight})
      : assert(child != null),
        this.controller = controller ?? ScrollableController(),
        this.minimumHeight = minimumHeight ?? 0.0,
        this.snapAbove = snapAbove ?? true,
        this.snapBelow = snapBelow ?? true,
        this.autoPop = autoPop ?? true,
        this.firstScrollTo = scrollTo ?? ScrollState.half,
        this.mayExceedChildHeight = mayExceedChildHeight ?? false;

  @override
  State<StatefulWidget> createState() => _ScrollableBottomSheetState();
}

class _ScrollableBottomSheetState extends State<ScrollableBottomSheet>
    with TickerProviderStateMixin
    implements ScrollableInterface {
  double _currentHeight;
  double _minimumHeight;
  double _halfHeight;
  bool _requestToFull = false;
  double _fullHeight;
  double _childHeight;
  ScrollController _scrollController = ScrollController();
  ScrollState _currentState;
  BuildContext _headerContext;
  AnimationController _activeAnimController;

  ScrollDirection _lastScrollDirection = ScrollDirection.none;

  @override
  void initState() {
    super.initState();
    widget.controller.interface = this;

    _minimumHeight = widget.minimumHeight ?? 0.0;
    _halfHeight = widget.halfHeight ?? 0.0;
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _childHeight = _headerContext.size.height;

      _currentState = widget.firstScrollTo;
      if (widget.callback != null) widget.callback(_currentState);
    });
  }

  @override
  void dispose() {
    _activeAnimController?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (_requestToFull) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        _animateTo(_fullHeight);
      });
      _requestToFull = false;
    }

    Widget child = LayoutBuilder(builder: (BuildContext context, BoxConstraints constraints) {
      if (widget.mayExceedChildHeight) {
        _fullHeight = constraints.maxHeight;
      } else {
        _fullHeight = math.min(_childHeight ?? _minimumHeight, constraints.maxHeight);
      }

      //for the first run
      if (_currentHeight == null) {
        _currentHeight = widget.firstScrollTo == ScrollState.half
            ? _halfHeight
            : widget.firstScrollTo == ScrollState.full ? _fullHeight : widget.minimumHeight;
      } else if (_currentHeight < widget.minimumHeight) {
        _currentHeight = widget.minimumHeight;
      }

      _currentHeight = _currentHeight.clamp(_minimumHeight, _fullHeight);

      return Container(
          height: _currentHeight,
          child: GestureDetector(
              onVerticalDragEnd: (DragEndDetails details) {
                double targetHeight;

                if (_scrollController.position.pixels < 0.0)
                  _scrollController.position.animateTo(0.0, duration: Duration(milliseconds: 200), curve: Curves.ease);

                if (_currentHeight <= _halfHeight) {
                  if (widget.snapBelow && _scrollController.hasClients && _scrollController.position.pixels <= 0) {
                    if (_lastScrollDirection == ScrollDirection.down) {
                      targetHeight = _minimumHeight;
                    } else {
                      targetHeight = _halfHeight;
                    }
                  }
                } else {
                  if (widget.snapAbove && _scrollController.hasClients && _scrollController.position.pixels <= 0) {
                    if (_lastScrollDirection == ScrollDirection.down) {
                      targetHeight = _halfHeight;
                    } else {
                      targetHeight = _fullHeight;
                    }
                  }
                }

                if (targetHeight != null) {
                  _animateTo(targetHeight, onComplete: () {
                    if ((targetHeight == 0.0 || targetHeight == _minimumHeight) && widget.autoPop) 
                    {
                        if(Navigator.of(context).canPop())      
                        Navigator.pop(context);
                    }
                    
                  });
                }

                _lastScrollDirection = ScrollDirection.none;

                if (_currentHeight >= _fullHeight && _currentState != ScrollState.full) {
                  _currentState = ScrollState.full;
                  if (widget.callback != null) widget.callback(_currentState);
                } else if (_currentState == ScrollState.full && _currentHeight < _halfHeight) {
                  _currentState = ScrollState.half;
                  if (widget.callback != null) widget.callback(_currentState);
                } else if (_currentState == ScrollState.minimum && _currentHeight >= _halfHeight) {
                  _currentState = ScrollState.half;
                  if (widget.callback != null) widget.callback(_currentState);
                } else if (_currentHeight <= widget.minimumHeight && _currentState != ScrollState.minimum) {
                  _currentState = ScrollState.minimum;
                  if (widget.callback != null) widget.callback(_currentState);
                }
                if (_currentHeight >= _fullHeight) {
                  _drag?.end(details);
                }
              },
              onVerticalDragDown: (DragDownDetails details) {
                _hold = _scrollController.position.hold(_disposeHold);
              },
              onVerticalDragCancel: () {
                _drag?.cancel();
                _hold?.cancel();
              },
              onVerticalDragUpdate: (DragUpdateDetails details) {
                if (details.primaryDelta > 0) {
                  // scroll downward
                  _lastScrollDirection = ScrollDirection.down;

                  if (_scrollController.offset <= 0.0) {
                    if (_currentHeight > 0.0) {
                      if (this.mounted) {
                        setState(() {
                          _currentHeight += details.primaryDelta * -1;
                        });
                      }
                    }
                  }
                } else if (details.primaryDelta < 0) {
                  // scroll upward
                  _lastScrollDirection = ScrollDirection.up;

                  if (_currentHeight < _fullHeight) {
                    if (this.mounted) {
                      setState(() {
                        _currentHeight += details.primaryDelta * -1;
                      });
                    }
                  }
                }

                if (_currentHeight >= _fullHeight) {
                  DragUpdateDetails d = details;

                  if (_scrollController.offset + -d.primaryDelta < 0.0) {
                    Offset newGlobalPosition = Offset(
                        details.globalPosition.dx, details.globalPosition.dy + -(d.primaryDelta) + _scrollController.offset);
                    Offset newDelta = Offset(details.delta.dx, _scrollController.offset);

                    d = DragUpdateDetails(
                        delta: newDelta,
                        primaryDelta: _scrollController.offset,
                        globalPosition: newGlobalPosition,
                        sourceTimeStamp: details.sourceTimeStamp);
                  }

                  _drag?.update(d);
                }
              },
              onVerticalDragStart: (DragStartDetails details) {
                if (_scrollController.position.maxScrollExtent > 0.0)
                  _drag = _scrollController.position.drag(details, _disposeDrag);
              },
              child: SingleChildScrollView(
                  controller: _scrollController,
                  physics: NeverScrollableScrollPhysics(),
                  child: Builder(builder: (BuildContext c) {
                    _headerContext = c;

                    return widget.child;
                  }))));
    });

    return child;
  }

  Drag _drag;
  ScrollHoldController _hold;

  void _disposeDrag() {
    _drag = null;
  }

  void _disposeHold() {
    _hold = null;
  }

  /// animate current Bottom Sheet to [minimumHeight] (if specified)
  /// you can call this method from Global Key
  ///
  /// See [example/main.dart] for example
  @override
  void animateToMinimum(BuildContext context, {bool willPop = false}) {
    if (widget.minimumHeight == null) return;

    _animateTo(_minimumHeight, onComplete: () {
      if (willPop) 
      {
         print("222222fffffffffffffffffffffffffffffffffffffffff"); 
         if(Navigator.of(context).canPop())
         Navigator.pop(context);
      }     
    });
  }

  /// animate current Bottom Sheet to 0.0
  /// you can call this method from Global Key
  ///
  /// See [example/main.dart] for example
  @override
  void animateToZero(BuildContext context, {bool willPop = false}) {
    _animateTo(0.0, onComplete: () {
      if (willPop) 
      {
        print("333333fffffffffffffffffffffffffffffffffffffffff");
        if(Navigator.of(context).canPop())
        Navigator.pop(context);
      }       
    });
  }

  /// animate current Bottom Sheet to initialHeight
  /// you can call this method from Global Key
  ///
  /// See [example/main.dart] for example
  @override
  void animateToHalf(BuildContext context) {
    _animateTo(_halfHeight);
  }

  /// animate current Bottom Sheet to maximum height available
  /// you can call this method from Global Key
  ///
  /// See [example/main.dart] for example
  @override
  void animateToFull(BuildContext context) {
    setState(() {
      _requestToFull = true;
    });
  }

  _animateTo(double targetHeight, {VoidCallback onComplete}) {
    if (!this.mounted) return;

    if (_scrollController.hasClients && _scrollController.position.pixels > 0) {
      _scrollController.animateTo(0.0, duration: Duration(milliseconds: 200), curve: Curves.ease);
    }

    AnimationController animationController = AnimationController(duration: const Duration(milliseconds: 200), vsync: this);

    _activeAnimController = animationController;

    var animation = CurvedAnimation(parent: animationController, curve: Interval(0.0, 1.0, curve: Curves.ease));

    Animation<double> zeroScale = Tween<double>(begin: _currentHeight, end: targetHeight).animate(animation);

    animationController.forward();

    var listener = () {
      if (this.mounted) {
        setState(() {
          _currentHeight = zeroScale.value;

          if (zeroScale.value == targetHeight && zeroScale.status == AnimationStatus.completed) {
            if (targetHeight == _fullHeight) {
              _currentState = ScrollState.full;
            } else if (targetHeight == _halfHeight) {
              _currentState = ScrollState.half;
            } else if (targetHeight == widget.minimumHeight) {
              _currentState = ScrollState.minimum;
            }

            if (onComplete != null) onComplete();
            animationController.dispose();
            _activeAnimController = null;

            if (widget.callback != null) widget.callback(_currentState);
          }
        });
      }
    };

    zeroScale.addListener(listener);
  }

  @override
  void setHalfHeight(double newHalfHeight) {
    _halfHeight = newHalfHeight;
  }

  @override
  void setMinimumHeight(double newMinimumHeight) {
    _minimumHeight = newMinimumHeight;
  }
}


example/lib/another_scrollable.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:scrollable_bottom_sheet/scrollable_bottom_sheet.dart';
import 'package:scrollable_bottom_sheet/scrollable_controller.dart';

class AnotherScrollable extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _AnotherScrollableState();
}

class _AnotherScrollableState extends State<AnotherScrollable> with TickerProviderStateMixin {
  final GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey<ScaffoldState>();
  final ScrollableController controller = ScrollableController();

  @override
  void initState() {
    super.initState();
    /*
    Uncomment this if you want to put bottom sheet into Route Stack,
    so that when you press back button, this bottom sheet will be closed

    WidgetsBinding.instance.addPostFrameCallback((_) {
      _scaffoldKey.currentState.showBottomSheet<void>(_bottomSheetBuilder);
    });
    */
  }

  List<Order> orders = [];
  List<String> randomFruits = [
    "Apple",
    "Apricot",
    "Blackberry",
    "Cherry",
    "Dragonfruit",
    "Grape",
    "Honeydew",
    "Jujube",
    "Kumquat",
    "Lime",
    "Papaya",
    "Passon Fruit",
    "Peach",
    "Pineapple",
    "Plum",
    "Raspberry",
    "Tomato"
  ];

  @override
  Widget build(BuildContext context) {
    math.Random random = math.Random();

    orders = List.generate(random.nextInt(100), (index) {
      return Order(randomFruits[random.nextInt(randomFruits.length - 1)], random.nextInt(100));
    });

    return Scaffold(
      bottomSheet: _bottomSheetBuilder(context),
      //if you use this bottomSheet, bottom Sheet will not put into Route Stack,
      // so when you press back button, it will pop this page
      key: _scaffoldKey,
      appBar: AppBar(
        title: Text("Scrollable by content"),
      ),
      body: Stack(
        children: <Widget>[
          Center(child: CircularProgressIndicator()),
          Container(
            constraints: BoxConstraints.expand(),
            // Add box decoration
            decoration: BoxDecoration(
              // Box decoration takes a gradient
              gradient: LinearGradient(
                // Where the linear gradient begins and ends
                begin: Alignment.topRight,
                end: Alignment.bottomLeft,
                // Add one stop for each color. Stops should increase from 0 to 1
                stops: [0.1, 0.5, 0.7, 0.9],
                colors: [
                  Color((math.Random().nextDouble() * 0xFFFFFF).toInt() << 0).withOpacity(1.0),
                  Color((math.Random().nextDouble() * 0xFFFFFF).toInt() << 0).withOpacity(1.0),
                  Color((math.Random().nextDouble() * 0xFFFFFF).toInt() << 0).withOpacity(1.0),
                  Color((math.Random().nextDouble() * 0xFFFFFF).toInt() << 0).withOpacity(1.0)
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _bottomSheetBuilder(BuildContext context) {
    return ScrollableBottomSheetByContent(
      _buildHeader(context),
      _buildContent(),
      autoPop: false,
      scrollTo: ScrollState.minimum,
      controller: controller,
    );
  }

 Widget _buildHeader(BuildContext context) {
    return Container(
        width: double.infinity,
        padding: EdgeInsets.all(16.0),
        color: Theme.of(context).primaryColor,
        child: Row(mainAxisSize: MainAxisSize.min, children: [
          Container(child: Icon(Icons.shopping_cart, color: Colors.white), margin: EdgeInsets.all(8.0)),
          Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
            Text("You've got order", style: Theme.of(context).primaryTextTheme.title),
            Text("Swipe up for details", style: Theme.of(context).primaryTextTheme.body1)
          ])
        ]));
  }

 Widget _buildContent() {
    int counter = 0;

    return Column(
        children: orders.map((order) {
      List<Widget> children = [];

      children.add(ListTile(title: Text(order.name), subtitle: Text("Qty: ${order.qty} pcs")));

      if (counter < orders.length) {
        children.add(Container(color: Colors.black38, height: 0.5));
      }
      return Column(children: children);
    }).toList());
  }
}

class Order {
  final String name;
  final int qty;

  Order(this.name, this.qty);
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:scrollable_bottom_sheet/scrollable_bottom_sheet.dart';
import 'package:scrollable_bottom_sheet/scrollable_controller.dart';
import 'package:scrollable_bottom_sheet_example/another_scrollable.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(home: BottomSheetDemo());
  }
}

class BottomSheetDemo extends StatefulWidget {
  @override
  _BottomSheetDemoState createState() => _BottomSheetDemoState();
}

class _BottomSheetDemoState extends State<BottomSheetDemo> with TickerProviderStateMixin {
  bool _bottomSheetActive = false;
  String _currentState = "initial";
  String _currentDirection = "up";
  final controller = ScrollableController();

  void _showMessage(BuildContext context) {
    showDialog<void>(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          content: const Text('You tapped the floating action button.'),
          actions: <Widget>[
            FlatButton(
                onPressed: () {
                  Navigator.pop(context);
                },
                child: const Text('OK'))
          ],
        );
      },
    );
  }

  final GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey<ScaffoldState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _scaffoldKey,
      appBar: AppBar(
        title: const Text('Persistent bottom sheet'),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _showMessage(context);
        },
        backgroundColor: Colors.redAccent,
        child: const Icon(
          Icons.add,
          semanticLabel: 'Add',
        ),
      ),
      body: Builder(
        builder: (BuildContext context) {
          return Column(children: [
            Expanded(
                child: Center(
                    child: RaisedButton(
                        onPressed: _bottomSheetActive
                            ? null
                            : () {
                                setState(() {
                                  //disable button
                                  _bottomSheetActive = true;
                                });
                                _showBottomSheet(context);
                              },
                        child: const Text('Show bottom sheet')))),
            Expanded(
                child: Center(
                    child: RaisedButton(
                        onPressed: () {
                          Navigator.push(context, MaterialPageRoute<Widget>(builder: (context) => AnotherScrollable()));
                        },
                        child: const Text('Show bottom sheet')))),
          ]);
        },
      ),
    );
  }

  Widget _bottomSheetBuilder(BuildContext context) {
    final ThemeData themeData = Theme.of(context);

    return Stack(children: [
      ScrollableBottomSheet(
        controller: controller,
        halfHeight: 250.0,
        minimumHeight: 50.0,
        autoPop: false,
        scrollTo: ScrollState.minimum,
        snapAbove: false,
        snapBelow: false,
        callback: (state) {
          if (state == ScrollState.minimum) {
            _currentState = "minimum";
            _currentDirection = "up";
          } else if (state == ScrollState.half) {
            if (_currentState == "minimum") {
              _currentDirection = "up";
            } else {
              _currentDirection = "down";
            }
            _currentState = "half";
          } else {
            _currentState = "full";
            _currentDirection = "down";
          }
        },
        child: Container(
            color: Colors.greenAccent,
            margin: EdgeInsets.only(bottom: 50.0),
            child: Padding(
                padding: const EdgeInsets.all(32.0),
                child: Column(children: [
                  InkWell(
                    child: Container(color: Colors.red, height: 57.0),
                    onTap: () {
                      controller.animateToZero(context, willPop: true);
                    },
                  ),
                  Text('This is a Material persistent bottom sheet. Drag downwards to dismiss it.',
                      textAlign: TextAlign.center, style: TextStyle(color: themeData.accentColor, fontSize: 24.0)),
                  Column(
                      children: List.generate(100, (index) {
                    return Text("Text $index");
                  }))
                ]))),
      ),
      Positioned(
          bottom: 0.0,
          left: 0.0,
          right: 0.0,
          height: 50.0,
          child: Material(
            elevation: 15.0,
            child: IconButton(
                icon: Icon(Icons.location_on),
                onPressed: () {
                  if (_currentState == "half") {
                    if (_currentDirection == "up") {
                      controller.animateToFull(context);
                    } else {
                      controller.animateToMinimum(context);
                    }
                  } else {
                    controller.animateToHalf(context);
                  }
                }),
          ))
    ]);
  }

 void _showBottomSheet(BuildContext context) {
    showBottomSheet<void>(context: context, builder: _bottomSheetBuilder).closed.whenComplete(() {
      if (mounted) {
        setState(() {
          // re-enable the button
          _bottomSheetActive = false;
        });
      }
    });
  }
}
